File: smart_campus/lib/services/fee_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/fee_model.dart';
import '../models/student_model.dart';
import 'auth_service.dart';

class FeeService {
  static final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  static const String _feeStructuresCollection = 'fee_structures';
  static const String _feeDuesCollection = 'fee_dues';
  static const String _paymentsCollection = 'payments';
  static const String _studentsCollection = 'students';

  // Mock data for development
  static List<FeeStructure> _mockFeeStructures = [];
  static List<FeeDue> _mockFeeDues = [];
  static List<Payment> _mockPayments = [];

  static void initializeMockData() {
    _mockFeeStructures = [
      FeeStructure(
        id: 'fee_struct_1',
        schoolId: 'school_1',
        name: 'Tuition Fee - Grade 5',
        description: 'Monthly tuition fee for Grade 5 students',
        type: FeeType.tuition,
        amount: 500.0,
        currency: 'USD',
        applicableClassId: 'class_1',
        applicableClassName: 'Grade 5A',
        isMandatory: true,
        isRecurring: true,
        recurringInterval: 1,
        dueDate: DateTime.now().add(const Duration(days: 15)),
        validFrom: DateTime.now().subtract(const Duration(days: 30)),
        validTo: DateTime.now().add(const Duration(days: 365)),
        isActive: true,
        createdAt: DateTime.now().subtract(const Duration(days: 30)),
        updatedAt: DateTime.now().subtract(const Duration(days: 5)),
      ),
      FeeStructure(
        id: 'fee_struct_2',
        schoolId: 'school_1',
        name: 'Transport Fee',
        description: 'Monthly transport fee for bus service',
        type: FeeType.transport,
        amount: 100.0,
        currency: 'USD',
        isMandatory: false,
        isRecurring: true,
        recurringInterval: 1,
        dueDate: DateTime.now().add(const Duration(days: 10)),
        validFrom: DateTime.now().subtract(const Duration(days: 30)),
        validTo: DateTime.now().add(const Duration(days: 365)),
        isActive: true,
        createdAt: DateTime.now().subtract(const Duration(days: 30)),
        updatedAt: DateTime.now().subtract(const Duration(days: 5)),
      ),
      FeeStructure(
        id: 'fee_struct_3',
        schoolId: 'school_1',
        name: 'Examination Fee',
        description: 'Annual examination fee',
        type: FeeType.examination,
        amount: 50.0,
        currency: 'USD',
        isMandatory: true,
        isRecurring: false,
        dueDate: DateTime.now().add(const Duration(days: 30)),
        validFrom: DateTime.now().subtract(const Duration(days: 30)),
        validTo: DateTime.now().add(const Duration(days: 365)),
        isActive: true,
        createdAt: DateTime.now().subtract(const Duration(days: 30)),
        updatedAt: DateTime.now().subtract(const Duration(days: 5)),
      ),
    ];

    _mockFeeDues = [
      FeeDue(
        id: 'fee_due_1',
        schoolId: 'school_1',
        studentId: 'student_1',
        studentName: 'Emma Smith',
        classId: 'class_1',
        className: 'Grade 5A',
        feeStructureId: 'fee_struct_1',
        feeName: 'Tuition Fee - Grade 5',
        feeType: FeeType.tuition,
        amount: 500.0,
        paidAmount: 500.0,
        balanceAmount: 0.0,
        currency: 'USD',
        dueDate: DateTime.now().add(const Duration(days: 15)),
        status: FeeStatus.paid,
        paidDate: DateTime.now().subtract(const Duration(days: 5)),
        payments: [],
        createdAt: DateTime.now().subtract(const Duration(days: 30)),
        updatedAt: DateTime.now().subtract(const Duration(days: 5)),
      ),
      FeeDue(
        id: 'fee_due_2',
        schoolId: 'school_1',
        studentId: 'student_1',
        studentName: 'Emma Smith',
        classId: 'class_1',
        className: 'Grade 5A',
        feeStructureId: 'fee_struct_2',
        feeName: 'Transport Fee',
        feeType: FeeType.transport,
        amount: 100.0,
        paidAmount: 0.0,
        balanceAmount: 100.0,
        currency: 'USD',
        dueDate: DateTime.now().add(const Duration(days: 10)),
        status: FeeStatus.pending,
        payments: [],
        createdAt: DateTime.now().subtract(const Duration(days: 30)),
        updatedAt: DateTime.now().subtract(const Duration(days: 5)),
      ),
      FeeDue(
        id: 'fee_due_3',
        schoolId: 'school_1',
        studentId: 'student_2',
        studentName: 'James Wilson',
        classId: 'class_2',
        className: 'Grade 6B',
        feeStructureId: 'fee_struct_1',
        feeName: 'Tuition Fee - Grade 5',
        feeType: FeeType.tuition,
        amount: 500.0,
        paidAmount: 300.0,
        balanceAmount: 200.0,
        currency: 'USD',
        dueDate: DateTime.now().subtract(const Duration(days: 5)),
        status: FeeStatus.overdue,
        payments: [],
        createdAt: DateTime.now().subtract(const Duration(days: 30)),
        updatedAt: DateTime.now().subtract(const Duration(days: 5)),
      ),
    ];

    _mockPayments = [
      Payment(
        id: 'payment_1',
        feeDueId: 'fee_due_1',
        amount: 500.0,
        currency: 'USD',
        method: PaymentMethod.bankTransfer,
        transactionId: 'TXN001',
        reference: 'Bank Transfer - Emma Smith Tuition',
        notes: 'Full payment received',
        paymentDate: DateTime.now().subtract(const Duration(days: 5)),
        recordedBy: 'admin_1',
        recordedByName: 'School Administrator',
        createdAt: DateTime.now().subtract(const Duration(days: 5)),
      ),
      Payment(
        id: 'payment_2',
        feeDueId: 'fee_due_3',
        amount: 300.0,
        currency: 'USD',
        method: PaymentMethod.cash,
        reference: 'Cash Payment - James Wilson',
        notes: 'Partial payment received',
        paymentDate: DateTime.now().subtract(const Duration(days: 10)),
        recordedBy: 'admin_1',
        recordedByName: 'School Administrator',
        createdAt: DateTime.now().subtract(const Duration(days: 10)),
      ),
    ];
  }

  // Create a new fee structure (School Admin only)
  static Future<String> createFeeStructure(FeeStructure feeStructure) async {
    try {
      final currentUser = AuthService.getCurrentUser();
      if (currentUser == null) throw Exception('User not authenticated');

      final docRef = _firestore.collection(_feeStructuresCollection).doc();
      final feeWithId = feeStructure.copyWith(
        id: docRef.id,
        createdAt: DateTime.now(),
        updatedAt: DateTime.now(),
      );

      await docRef.set(feeWithId.toMap());
      return docRef.id;
    } catch (e) {
      print('Error creating fee structure: $e');
      // For development, add to mock data
      final id = 'fee_struct_${DateTime.now().millisecondsSinceEpoch}';
      _mockFeeStructures.add(feeStructure.copyWith(id: id));
      return id;
    }
  }

  // Get fee structures by school ID
  static Future<List<FeeStructure>> getFeeStructuresBySchool(String schoolId) async {
    try {
      final currentUser = AuthService.getCurrentUser();
      if (currentUser == null) throw Exception('User not authenticated');

      final querySnapshot = await _firestore
          .collection(_feeStructuresCollection)
          .where('schoolId', isEqualTo: schoolId)
          .where('isActive', isEqualTo: true)
          .orderBy('createdAt', descending: true)
          .get();

      return querySnapshot.docs
          .map((doc) => FeeStructure.fromMap(doc.data()))
          .toList();
    } catch (e) {
      print('Error getting fee structures: $e');
      // Return mock data for development
      return _mockFeeStructures
          .where((fee) => fee.schoolId == schoolId && fee.isActive)
          .toList();
    }
  }

  // Get fee dues by school ID
  static Future<List<FeeDue>> getFeeDuesBySchool(String schoolId) async {
    try {
      final currentUser = AuthService.getCurrentUser();
      if (currentUser == null) throw Exception('User not authenticated');

      final querySnapshot = await _firestore
          .collection(_feeDuesCollection)
          .where('schoolId', isEqualTo: schoolId)
          .orderBy('createdAt', descending: true)
          .get();

      return querySnapshot.docs
          .map((doc) => FeeDue.fromMap(doc.data()))
          .toList();
    } catch (e) {
      print('Error getting fee dues: $e');
      // Return mock data for development
      return _mockFeeDues
          .where((due) => due.schoolId == schoolId)
          .toList();
    }
  }

  // Get fee dues by student ID (for parents)
  static Future<List<FeeDue>> getFeeDuesByStudent(String studentId) async {
    try {
      final currentUser = AuthService.getCurrentUser();
      if (currentUser == null) throw Exception('User not authenticated');

      final querySnapshot = await _firestore
          .collection(_feeDuesCollection)
          .where('studentId', isEqualTo: studentId)
          .orderBy('createdAt', descending: true)
          .get();

      return querySnapshot.docs
          .map((doc) => FeeDue.fromMap(doc.data()))
          .toList();
    } catch (e) {
      print('Error getting fee dues by student: $e');
      // Return mock data for development
      return _mockFeeDues
          .where((due) => due.studentId == studentId)
          .toList();
    }
  }

  // Update fee due status (School Admin only)
  static Future<void> updateFeeDueStatus(
    String feeDueId,
    FeeStatus status, {
    double? paidAmount,
    String? notes,
  }) async {
    try {
      final currentUser = AuthService.getCurrentUser();
      if (currentUser == null) throw Exception('User not authenticated');

      // Get current fee due
      final feeDue = _mockFeeDues.firstWhere((due) => due.id == feeDueId);
      final newBalanceAmount = feeDue.amount - (paidAmount ?? feeDue.paidAmount);
      
      final updateData = {
        'status': status.name,
        'paidAmount': paidAmount ?? feeDue.paidAmount,
        'balanceAmount': newBalanceAmount,
        'paidDate': status == FeeStatus.paid ? Timestamp.fromDate(DateTime.now()) : feeDue.paidDate,
        'notes': notes ?? feeDue.notes,
        'updatedAt': Timestamp.fromDate(DateTime.now()),
      };

      await _firestore.collection(_feeDuesCollection).doc(feeDueId).update(updateData);
      
      // Update mock data for development
      final index = _mockFeeDues.indexWhere((due) => due.id == feeDueId);
      if (index != -1) {
        _mockFeeDues[index] = _mockFeeDues[index].copyWith(
          status: status,
          paidAmount: paidAmount ?? _mockFeeDues[index].paidAmount,
          balanceAmount: newBalanceAmount,
          paidDate: status == FeeStatus.paid ? DateTime.now() : _mockFeeDues[index].paidDate,
          notes: notes ?? _mockFeeDues[index].notes,
          updatedAt: DateTime.now(),
        );
      }
    } catch (e) {
      print('Error updating fee due status: $e');
      // Update mock data for development
      final index = _mockFeeDues.indexWhere((due) => due.id == feeDueId);
      if (index != -1) {
        final paidAmountValue = paidAmount ?? _mockFeeDues[index].paidAmount;
        final newBalanceAmount = _mockFeeDues[index].amount - paidAmountValue;
        
        _mockFeeDues[index] = _mockFeeDues[index].copyWith(
          status: status,
          paidAmount: paidAmountValue,
          balanceAmount: newBalanceAmount,
          paidDate: status == FeeStatus.paid ? DateTime.now() : _mockFeeDues[index].paidDate,
          notes: notes ?? _mockFeeDues[index].notes,
          updatedAt: DateTime.now(),
        );
      }
    }
  }

  // Record a payment (School Admin only)
  static Future<String> recordPayment(Payment payment) async {
    try {
      final currentUser = AuthService.getCurrentUser();
      if (currentUser == null) throw Exception('User not authenticated');

      final docRef = _firestore.collection(_paymentsCollection).doc();
      final paymentWithId = Payment(
        id: docRef.id,
        feeDueId: payment.feeDueId,
        amount: payment.amount,
        currency: payment.currency,
        method: payment.method,
        transactionId: payment.transactionId,
        reference: payment.reference,
        notes: payment.notes,
        paymentDate: payment.paymentDate,
        recordedBy: currentUser.uid,
        recordedByName: currentUser.displayName ?? 'Admin User',
        createdAt: DateTime.now(),
      );

      await docRef.set(paymentWithId.toMap());
      
      // Update fee due
      final feeDue = _mockFeeDues.firstWhere((due) => due.id == payment.feeDueId);
      final newPaidAmount = feeDue.paidAmount + payment.amount;
      final newBalanceAmount = feeDue.amount - newPaidAmount;
      final newStatus = newBalanceAmount <= 0 ? FeeStatus.paid : FeeStatus.partial;
      
      await updateFeeDueStatus(
        payment.feeDueId,
        newStatus,
        paidAmount: newPaidAmount,
      );
      
      return docRef.id;
    } catch (e) {
      print('Error recording payment: $e');
      // For development, add to mock data
      final id = 'payment_${DateTime.now().millisecondsSinceEpoch}';
      _mockPayments.add(payment.copyWith(id: id));
      return id;
    }
  }

  // Get overdue fees (for notifications)
  static Future<List<FeeDue>> getOverdueFees(String schoolId) async {
    try {
      final currentUser = AuthService.getCurrentUser();
      if (currentUser == null) throw Exception('User not authenticated');

      final now = DateTime.now();
      final querySnapshot = await _firestore
          .collection(_feeDuesCollection)
          .where('schoolId', isEqualTo: schoolId)
          .where('status', whereIn: [FeeStatus.pending.name, FeeStatus.partial.name])
          .where('dueDate', isLessThan: Timestamp.fromDate(now))
          .get();

      return querySnapshot.docs
          .map((doc) => FeeDue.fromMap(doc.data()))
          .toList();
    } catch (e) {
      print('Error getting overdue fees: $e');
      // Return mock data for development
      final now = DateTime.now();
      return _mockFeeDues
          .where((due) => due.schoolId == schoolId && 
                         (due.status == FeeStatus.pending || due.status == FeeStatus.partial) &&
                         due.dueDate.isBefore(now))
          .toList();
    }
  }

  // Get fee statistics for school admin
  static Future<Map<String, dynamic>> getFeeStatistics(String schoolId) async {
    try {
      final feeDues = await getFeeDuesBySchool(schoolId);
      
      double totalAmount = 0;
      double totalPaid = 0;
      double totalPending = 0;
      double totalOverdue = 0;
      int paidCount = 0;
      int pendingCount = 0;
      int overdueCount = 0;
      
      for (final due in feeDues) {
        totalAmount += due.amount;
        totalPaid += due.paidAmount;
        
        switch (due.status) {
          case FeeStatus.paid:
            paidCount++;
            break;
          case FeeStatus.pending:
            pendingCount++;
            totalPending += due.balanceAmount;
            break;
          case FeeStatus.overdue:
            overdueCount++;
            totalOverdue += due.balanceAmount;
            break;
          case FeeStatus.partial:
            pendingCount++;
            totalPending += due.balanceAmount;
            break;
          default:
            break;
        }
      }
      
      return {
        'totalAmount': totalAmount,
        'totalPaid': totalPaid,
        'totalPending': totalPending,
        'totalOverdue': totalOverdue,
        'paidCount': paidCount,
        'pendingCount': pendingCount,
        'overdueCount': overdueCount,
        'totalCount': feeDues.length,
        'collectionRate': totalAmount > 0 ? (totalPaid / totalAmount) * 100 : 0,
      };
    } catch (e) {
      print('Error getting fee statistics: $e');
      return {
        'totalAmount': 0.0,
        'totalPaid': 0.0,
        'totalPending': 0.0,
        'totalOverdue': 0.0,
        'paidCount': 0,
        'pendingCount': 0,
        'overdueCount': 0,
        'totalCount': 0,
        'collectionRate': 0.0,
      };
    }
  }

  // Get students for fee management
  static Future<List<Student>> getStudentsForFeeManagement(String schoolId) async {
    try {
      final currentUser = AuthService.getCurrentUser();
      if (currentUser == null) throw Exception('User not authenticated');

      final querySnapshot = await _firestore
          .collection(_studentsCollection)
          .where('schoolId', isEqualTo: schoolId)
          .orderBy('name')
          .get();

      return querySnapshot.docs
          .map((doc) => Student.fromMap(doc.data()))
          .toList();
    } catch (e) {
      print('Error getting students for fee management: $e');
      // Return mock data for development
      return [
        Student(
          id: 'student_1',
          name: 'Emma Smith',
          email: 'emma.smith@school.com',
          className: 'Grade 5A',
          rollNo: 'ST001',
          phone: '+1234567890',
          address: '123 Main St, Downtown',
          parentName: 'John Smith',
          parentPhone: '+0987654321',
          dateOfBirth: DateTime(2015, 6, 15),
          gender: 'Female',
          bloodGroup: 'O+',
          admissionDate: '2020-01-01',
          // emergencyContact: 'John Smith - +1234567890', // Not in Student model
          // medicalInfo: 'No known allergies', // Not in Student model
          // createdAt: DateTime.now(), // Not in Student model
          // updatedAt: DateTime.now(), // Not in Student model
        ),
        Student(
          id: 'student_2',
          name: 'James Wilson',
          email: 'james.wilson@school.com',
          className: 'Grade 6B',
          rollNo: 'ST002',
          phone: '+1234567891',
          address: '456 Oak St, Downtown',
          parentName: 'Sarah Wilson',
          parentPhone: '+0987654322',
          dateOfBirth: DateTime(2014, 3, 20),
          gender: 'Male',
          bloodGroup: 'A+',
          admissionDate: '2020-01-01',
          // emergencyContact: 'Sarah Wilson - +1234567891', // Not in Student model
          // medicalInfo: 'Asthma', // Not in Student model
          // createdAt: DateTime.now(), // Not in Student model
          // updatedAt: DateTime.now(), // Not in Student model
        ),
      ];
    }
  }

  // Delete fee structure
  static Future<void> deleteFeeStructure(String feeStructureId) async {
    try {
      final currentUser = AuthService.getCurrentUser();
      if (currentUser == null) throw Exception('User not authenticated');

      await _firestore.collection(_feeStructuresCollection).doc(feeStructureId).delete();
    } catch (e) {
      print('Error deleting fee structure: $e');
      // Remove from mock data for development
      _mockFeeStructures.removeWhere((fee) => fee.id == feeStructureId);
    }
  }

  // Update fee structure
  static Future<void> updateFeeStructure(FeeStructure feeStructure) async {
    try {
      final currentUser = AuthService.getCurrentUser();
      if (currentUser == null) throw Exception('User not authenticated');

      await _firestore.collection(_feeStructuresCollection).doc(feeStructure.id).update({
        ...feeStructure.toMap(),
        'updatedAt': Timestamp.fromDate(DateTime.now()),
      });
    } catch (e) {
      print('Error updating fee structure: $e');
      // Update mock data for development
      final index = _mockFeeStructures.indexWhere((fee) => fee.id == feeStructure.id);
      if (index != -1) {
        _mockFeeStructures[index] = feeStructure.copyWith(updatedAt: DateTime.now());
      }
    }
  }
}

// Extension for Payment to add copyWith method
extension PaymentCopyWith on Payment {
  Payment copyWith({
    String? id,
    String? feeDueId,
    double? amount,
    String? currency,
    PaymentMethod? method,
    String? transactionId,
    String? reference,
    String? notes,
    DateTime? paymentDate,
    String? recordedBy,
    String? recordedByName,
    DateTime? createdAt,
  }) {
    return Payment(
      id: id ?? this.id,
      feeDueId: feeDueId ?? this.feeDueId,
      amount: amount ?? this.amount,
      currency: currency ?? this.currency,
      method: method ?? this.method,
      transactionId: transactionId ?? this.transactionId,
      reference: reference ?? this.reference,
      notes: notes ?? this.notes,
      paymentDate: paymentDate ?? this.paymentDate,
      recordedBy: recordedBy ?? this.recordedBy,
      recordedByName: recordedByName ?? this.recordedByName,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}

File: SmartCampusMobile/services/FeeService.ts
import { FeeStructure, Payment, FeeStats, FeeItem, FeeStatus, PaymentStatus } from '../models/FeeModel';

export class FeeService {
  private static instance: FeeService;
  private feeData: FeeStructure[] = [];
  private paymentData: Payment[] = [];

  static getInstance(): FeeService {
    if (!FeeService.instance) {
      FeeService.instance = new FeeService();
      FeeService.instance.initializeMockData();
    }
    return FeeService.instance;
  }

  // Initialize mock data
  initializeMockData(): void {
    this.feeData = [
      {
        id: 'fee_1',
        studentId: 'student_1',
        studentName: 'Emma Johnson',
        classId: 'class_1',
        className: 'Grade 5A',
        academicYear: '2024-25',
        fees: [
          {
            id: 'fee_item_1',
            name: 'Tuition Fee',
            type: 'tuition',
            amount: 5000,
            dueDate: new Date('2025-01-15'),
            status: 'paid',
            description: 'Monthly tuition fee',
            isMandatory: true,
          },
          {
            id: 'fee_item_2',
            name: 'Transport Fee',
            type: 'transport',
            amount: 800,
            dueDate: new Date('2025-01-15'),
            status: 'paid',
            description: 'Monthly transport fee',
            isMandatory: true,
          },
          {
            id: 'fee_item_3',
            name: 'Library Fee',
            type: 'library',
            amount: 200,
            dueDate: new Date('2025-01-20'),
            status: 'pending',
            description: 'Annual library membership',
            isMandatory: false,
          },
        ],
        totalAmount: 6000,
        paidAmount: 5800,
        balanceAmount: 200,
        status: 'partial',
        createdAt: new Date('2025-01-01'),
        updatedAt: new Date('2025-01-15'),
      },
      {
        id: 'fee_2',
        studentId: 'student_2',
        studentName: 'James Wilson',
        classId: 'class_1',
        className: 'Grade 5A',
        academicYear: '2024-25',
        fees: [
          {
            id: 'fee_item_4',
            name: 'Tuition Fee',
            type: 'tuition',
            amount: 5000,
            dueDate: new Date('2025-01-15'),
            status: 'overdue',
            description: 'Monthly tuition fee',
            isMandatory: true,
          },
          {
            id: 'fee_item_5',
            name: 'Sports Fee',
            type: 'sports',
            amount: 300,
            dueDate: new Date('2025-01-10'),
            status: 'overdue',
            description: 'Annual sports fee',
            isMandatory: true,
          },
        ],
        totalAmount: 5300,
        paidAmount: 0,
        balanceAmount: 5300,
        status: 'overdue',
        createdAt: new Date('2025-01-01'),
        updatedAt: new Date('2025-01-15'),
      },
      {
        id: 'fee_3',
        studentId: 'student_3',
        studentName: 'Sarah Davis',
        classId: 'class_2',
        className: 'Grade 5B',
        academicYear: '2024-25',
        fees: [
          {
            id: 'fee_item_6',
            name: 'Tuition Fee',
            type: 'tuition',
            amount: 5000,
            dueDate: new Date('2025-01-15'),
            status: 'paid',
            description: 'Monthly tuition fee',
            isMandatory: true,
          },
          {
            id: 'fee_item_7',
            name: 'Exam Fee',
            type: 'exam',
            amount: 500,
            dueDate: new Date('2025-01-25'),
            status: 'pending',
            description: 'Mid-term exam fee',
            isMandatory: true,
          },
        ],
        totalAmount: 5500,
        paidAmount: 5000,
        balanceAmount: 500,
        status: 'partial',
        createdAt: new Date('2025-01-01'),
        updatedAt: new Date('2025-01-15'),
      },
    ];

    this.paymentData = [
      {
        id: 'pay_1',
        feeStructureId: 'fee_1',
        studentId: 'student_1',
        amount: 5800,
        paymentMethod: 'bank_transfer',
        paymentDate: new Date('2025-01-15'),
        transactionId: 'TXN001',
        receiptNumber: 'RCP001',
        status: 'completed',
        notes: 'Payment for tuition and transport',
        processedBy: 'admin_1',
        createdAt: new Date('2025-01-15'),
      },
      {
        id: 'pay_2',
        feeStructureId: 'fee_3',
        studentId: 'student_3',
        amount: 5000,
        paymentMethod: 'card',
        paymentDate: new Date('2025-01-14'),
        transactionId: 'TXN002',
        receiptNumber: 'RCP002',
        status: 'completed',
        notes: 'Tuition fee payment',
        processedBy: 'admin_1',
        createdAt: new Date('2025-01-14'),
      },
    ];
  }

  // Get fee structure for a student
  async getFeeStructureForStudent(studentId: string): Promise<FeeStructure | null> {
    return new Promise((resolve) => {
      setTimeout(() => {
        const feeStructure = this.feeData.find(fee => fee.studentId === studentId);
        resolve(feeStructure || null);
      }, 400);
    });
  }

  // Get fee structures for a class
  async getFeeStructuresForClass(classId: string): Promise<FeeStructure[]> {
    return new Promise((resolve) => {
      setTimeout(() => {
        const feeStructures = this.feeData.filter(fee => fee.classId === classId);
        resolve(feeStructures);
      }, 400);
    });
  }

  // Get all fee structures (for admin)
  async getAllFeeStructures(): Promise<FeeStructure[]> {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve(this.feeData);
      }, 300);
    });
  }

  // Record a payment
  async recordPayment(payment: Omit<Payment, 'id' | 'createdAt'>): Promise<{ success: boolean; payment?: Payment; message: string }> {
    return new Promise((resolve) => {
      setTimeout(() => {
        try {
          const newPayment: Payment = {
            ...payment,
            id: `pay_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
            createdAt: new Date(),
          };

          this.paymentData.push(newPayment);

          // Update fee structure
          const feeStructure = this.feeData.find(fee => fee.id === payment.feeStructureId);
          if (feeStructure) {
            feeStructure.paidAmount += payment.amount;
            feeStructure.balanceAmount = feeStructure.totalAmount - feeStructure.paidAmount;
            
            // Update fee status
            if (feeStructure.balanceAmount <= 0) {
              feeStructure.status = 'paid';
              // Update individual fee items
              feeStructure.fees.forEach(fee => {
                if (fee.status === 'pending' || fee.status === 'overdue') {
                  fee.status = 'paid';
                }
              });
            } else {
              feeStructure.status = 'partial';
            }
            
            feeStructure.updatedAt = new Date();
          }

          resolve({ success: true, payment: newPayment, message: 'Payment recorded successfully' });
        } catch (error) {
          resolve({ success: false, message: 'Failed to record payment' });
        }
      }, 600);
    });
  }

  // Get payment history for a student
  async getPaymentHistory(studentId: string): Promise<Payment[]> {
    return new Promise((resolve) => {
      setTimeout(() => {
        const payments = this.paymentData.filter(payment => payment.studentId === studentId);
        resolve(payments.sort((a, b) => b.paymentDate.getTime() - a.paymentDate.getTime()));
      }, 300);
    });
  }

  // Get fee statistics
  async getFeeStats(): Promise<FeeStats> {
    return new Promise((resolve) => {
      setTimeout(() => {
        const totalFees = this.feeData.reduce((sum, fee) => sum + fee.totalAmount, 0);
        const collectedFees = this.feeData.reduce((sum, fee) => sum + fee.paidAmount, 0);
        const pendingFees = this.feeData.filter(fee => fee.status === 'pending').length;
        const overdueFees = this.feeData.filter(fee => fee.status === 'overdue').length;
        const collectionRate = totalFees > 0 ? Math.round((collectedFees / totalFees) * 100) : 0;
        
        // Calculate monthly collection (current month)
        const currentMonth = new Date().getMonth();
        const currentYear = new Date().getFullYear();
        const monthlyCollection = this.paymentData
          .filter(payment => 
            payment.paymentDate.getMonth() === currentMonth && 
            payment.paymentDate.getFullYear() === currentYear &&
            payment.status === 'completed'
          )
          .reduce((sum, payment) => sum + payment.amount, 0);

        resolve({
          totalFees,
          collectedFees,
          pendingFees,
          overdueFees,
          collectionRate,
          monthlyCollection,
        });
      }, 400);
    });
  }

  // Update fee status
  async updateFeeStatus(feeStructureId: string, feeItemId: string, status: FeeStatus): Promise<{ success: boolean; message: string }> {
    return new Promise((resolve) => {
      setTimeout(() => {
        try {
          const feeStructure = this.feeData.find(fee => fee.id === feeStructureId);
          if (feeStructure) {
            const feeItem = feeStructure.fees.find(fee => fee.id === feeItemId);
            if (feeItem) {
              feeItem.status = status;
              feeStructure.updatedAt = new Date();
              resolve({ success: true, message: 'Fee status updated successfully' });
            } else {
              resolve({ success: false, message: 'Fee item not found' });
            }
          } else {
            resolve({ success: false, message: 'Fee structure not found' });
          }
        } catch (error) {
          resolve({ success: false, message: 'Failed to update fee status' });
        }
      }, 400);
    });
  }

  // Generate fee reminder
  async generateFeeReminder(studentId: string): Promise<{ success: boolean; message: string }> {
    return new Promise((resolve) => {
      setTimeout(() => {
        // Mock fee reminder generation
        console.log(`Fee reminder generated for student ${studentId}`);
        resolve({ success: true, message: 'Fee reminder generated successfully' });
      }, 300);
    });
  }

  // Create fee structure
  async createFeeStructure(feeStructure: Omit<FeeStructure, 'id' | 'createdAt' | 'updatedAt'>): Promise<{ success: boolean; feeStructure?: FeeStructure; message: string }> {
    return new Promise((resolve) => {
      setTimeout(() => {
        try {
          const newFeeStructure: FeeStructure = {
            ...feeStructure,
            id: `fee_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
            createdAt: new Date(),
            updatedAt: new Date(),
          };

          this.feeData.push(newFeeStructure);
          resolve({ success: true, feeStructure: newFeeStructure, message: 'Fee structure created successfully' });
        } catch (error) {
          resolve({ success: false, message: 'Failed to create fee structure' });
        }
      }, 600);
    });
  }
}

export default FeeService.getInstance();